### DISCUSSION TITLE
Is UHCI driver in BleskOS polling-driven or interrupt-driven?

### DISCUSSION THREAD

**Discussion by davidgaosheng** [VendelinSlezak/BleskOS#19] [Q&A] [answered]
does the driver send an interrupt signal to host CPU when it receives a character from keyboard?

**Comment by VendelinSlezak** [comment, marked as answer]
Well, BleskOS uses quite unique approach. Bulk transfers are pooled. However, interrupt transfers are neither pooled, or driven by direct interrupts. Instead, checking for UHCI interrupt transfer is done in processes in background, so it is driven by PIT interrupt. When UHCI driver recognizes USB mouse or USB keyboard, it sets variabile, that every time processes in background are called by PIT interrupt, it check if some transfer happen. If yes, it process it. Advantage if this approach is that we do not need to do any hard work with remapping interrupts or deal with shared interrupts, we use only one interrupt for USB mouse and USB keyboard. And because fastest interrupt transfer allowed is 4 ms and PIT interrupt is fired every 2 ms, there will never be missed packet, all packets will be processed.

You can check code here: https://github.com/VendelinSlezak/BleskOS/blob/master/source/drivers/system/pit.c#L23 https://github.com/VendelinSlezak/BleskOS/blob/master/source/drivers/system/processes_on_background.c#L109
